synapse/federation/sender/__init__.py

```diff
@@ -393,34 +420,59 @@ async def handle_room_events(events: Iterable[EventBase]) -> None:
         finally:
             self._is_processing = False
 
-    async def _send_pdu(self, pdu: EventBase, destinations: Iterable[str]) -> None:
-        # We loop through all destinations to see whether we already have
-        # a transaction in progress. If we do, stick it in the pending_pdus
-        # table and we'll get back to it later.
+    async def _distribute_events(
+        self,
+        per_room_events_and_dests: List[
+            Tuple[str, List[Tuple[EventBase, Collection[str]]]]
+        ],
+    ) -> None:
+        # Tuples of room_id + destination to their max-seen stream_ordering
+        room_with_dest_stream_ordering = {}  # type: Dict[Tuple[str, str], int]
 
-        destinations = set(destinations)
-        destinations.discard(self.server_name)
-        logger.debug("Sending to: %s", str(destinations))
+        # List of events to send to each destination
+        events_by_dest = {}  # type: Dict[str, List[EventBase]]
 
-        if not destinations:
-            return
+        # For each roomID + events with destinations pair...
+        for room_id, event_and_dests in per_room_events_and_dests:
+
+            # (dest -> stream_id)
+            this_stream_ordering = {}  # type: Dict[str, int]
+
+            # ...get every event with their destinations...
+            for event, destinations in event_and_dests:
 
-        sent_pdus_destination_dist_total.inc(len(destinations))
-        sent_pdus_destination_dist_count.inc()
+                # (we got this from the database, it's filled)
+                assert event.internal_metadata.stream_ordering
 
-        assert pdu.internal_metadata.stream_ordering
+                sent_pdus_destination_dist_total.inc(len(destinations))
+                sent_pdus_destination_dist_count.inc()
 
-        # track the fact that we have a PDU for these destinations,
-        # to allow us to perform catch-up later on if the remote is unreachable
-        # for a while.
-        await self.store.store_destination_rooms_entries(
-            destinations,
-            pdu.room_id,
-            pdu.internal_metadata.stream_ordering,
+                # ...iterate over those destinations..
+                for destination in destinations:
+                    if (
+                        this_stream_ordering.setdefault(destination, 0)
+                        < event.internal_metadata.stream_ordering
+                    ):
+                        # ...update their stream-ordering...
+                        this_stream_ordering[
+                            destination
+                        ] = event.internal_metadata.stream_ordering
+
+                    # ...and add the event to each destination queue.
+                    events_by_dest.setdefault(destination, []).append(event)
+
+            for dest, stream_id in this_stream_ordering.items():
+                room_with_dest_stream_ordering[(room_id, dest)] = stream_id
```

Comment:
why not update `room_with_dest_stream_ordering` directly, rather than going via `this_stream_ordering`?

Reply:
You suggested changing *from* that approach to this approach earlier [here](https://github.com/matrix-org/synapse/pull/9702#discussion_r605109179), though i can see how we could make this `dict[str, dict[str, int]]` instead, i'll do that and update some storage logic :+1:

Reply:
> You suggested changing from that approach to this approach earlier here

no, that's not what I suggested there.

Reply:
then what did you suggest?

Reply:
> you could omit `room_id` from `per_room_events_and_dests`

it seems fairly clear to me, but to spell it out:

currently, `per_room_events_and_dests` is a list of `(room_id, [(event, destinations), ...])` tuples. I assert that you don't need to keep `room_id` separately there, and could instead just make it a it a list of lists of `(event, destinations)`.

I think doing so would simplify the implementation of `get_federatable_events_and_destinations` (since it would no longer need a separate `room_id` parameter).

Reply:
Alright, that makes sense, i'll try to change it to this :+1: